### PR TITLE
MOB-1878 ci release candidate workflow #2

### PR DESCRIPTION
**Pull request by d4r1091**
[MOB-1878](https://ecosia.atlassian.net/browse/MOB-1878)

## Context

We iterated on the first improvement attempt https://github.com/ecosia/ios-browser/pull/498. After some sharing and 🧠 🌪️ we ended up keeping the structure essentially the same as it is now but moved over the `Common.xcconfig`. The *Approach* section will describe the new job.
The iteration of new versions (RCs) will be based on patches.
Each new release will be a new minor version (unless specific need of having a major).
The `CURRENT_PROJECT_VERSION` is set to 0. Indicating that is not being used for local testing as it will get updated by the CircleCI run.
Updated README as well.

### Job description

The `build-testflight-deploy` job within the CircleCI configuration `deploy-release.yml` has mutated.
Here is a more detailed description of the job:

1. Environment Setup:
   - The job sets the environment variable `CI` to `true`, indicating that it is running in a continuous integration environment.
   - It specifies the macOS environment with Xcode version 14.3.1.
   - The job is allocated the `macos.x86.medium.gen2` resource class, which determines the hardware and performance characteristics for the job execution.

2. Steps:
   a. Checkout:
      - This step checks out the code from the version control system.

   b. Check for MARKETING_VERSION change ( ⭐ ⭐ ⭐  important ⭐ ⭐ ⭐ ) :
      - The job fetches the latest code for the specified branch and checks it out.
      - It extracts the current `MARKETING_VERSION` value from the `Client/Configuration/Common.xcconfig` file.
      - Then, it checks out the previous commit and retrieves the old `MARKETING_VERSION` value.
      - If the current and old `MARKETING_VERSION` values are the same, the job prints a message indicating that the version has not changed and halts (gracefully see: [CirleCI attempt](https://app.circleci.com/pipelines/github/ecosia/ios-browser/2436/workflows/686cffcd-f0b9-4af4-808d-1c9f807fb547)) the execution of the step.

   c. Restore Cache:
      - This step restores cached dependencies (Gems) based on the key generated from the `Gemfile.lock` checksum.

   d. Bundle Check or Install:
      - The job checks if the bundle is already installed, and if not, it installs the necessary dependencies using the `bundle install` command with the `--path vendor/bundle` option.

   e. Save Cache:
      - The step saves the dependencies cache using the key generated from the `Gemfile.lock` checksum.

   f. Verify GitHub Host:
      - This step verifies the GitHub host by retrieving SSH fingerprints for GitHub's IP addresses and adding them to the `~/.ssh/known_hosts` file.

   g. Clean Packages:
      - The job cleans the Xcode project by executing the `clean` command for the `Ecosia` scheme in the `Client.xcodeproj`.

   h. Setup Content Blocking Scripts:
      - This step sets up content blocking scripts by executing the `content_blocker_update.sh` script.

   i. Build and Deploy to Testflight:
      - The job builds and deploys the iOS application to Testflight using Fastlane.
      - It executes the `testflight_live` lane from Fastlane, leveraging the `bundle exec` command.

### Lane description

The Fastlane lane `testflight_live` performs the following actions:

1. `version_number = get_xcconfig_value(path: 'Client/Configuration/Common.xcconfig', name: 'MARKETING_VERSION')`: This line retrieves the value of the `MARKETING_VERSION` key from the `Common.xcconfig` file located at the path `Client/Configuration/Common.xcconfig`. The value is stored in the `version_number` variable.

2. `ci_run_build_number = ENV["CIRCLE_BUILD_NUM"]`: This line retrieves the value of the `CIRCLE_BUILD_NUM` environment variable and stores it in the `ci_run_build_number` variable. This variable likely holds the build number associated with the continuous integration (CI) system being used.

3. `udpated_build_number = increment_build_number(build_number: ci_run_build_number)`: This line increments the `ci_run_build_number` by one using the `increment_build_number` action, and the result is stored in the `updated_build_number` variable.

4. `match(app_identifier: ['com.ecosia.ecosiaapp', 'com.ecosia.ecosiaapp.WidgetKit', 'com.ecosia.ecosiaapp.ShareTo'], type: 'appstore', readonly: true)`: This line utilizes the `match` action to retrieve the necessary certificates and provisioning profiles for the specified app identifiers. It fetches the required credentials for app identifiers `'com.ecosia.ecosiaapp'`, `'com.ecosia.ecosiaapp.WidgetKit'`, and `'com.ecosia.ecosiaapp.ShareTo'` from the App Store. The `readonly` option indicates that the existing credentials should be used without modifying them.

5. `gym(scheme: "Ecosia", export_method: "app-store", configuration: "Release")`: This line triggers the `gym` action, which builds the app using the specified scheme (`Ecosia`), export method (`app-store`), and configuration (`Release`).

6. `pilot(skip_waiting_for_build_processing: true, apple_id: "670881887", itc_provider: "33YMRSYD2L")`: This line uses the `pilot` action to upload the built app to TestFlight for beta testing and distribution. The `skip_waiting_for_build_processing` option allows skipping the wait for the build to process before distributing it. The `apple_id` and `itc_provider` options specify the Apple ID and iTunes Connect provider short name, respectively.

7. `tag_and_push(version_number: version_number, build_number: updated_build_number)`: This line utilizes the `tag_and_push` action to create a Git tag with the `version_number` and `build_number` as parameters, and then pushes the tag to the remote repository.

## Other

After some thinking process, I've realized that the most efficient (CI + complexity wise) approach would be to keep the `path-filtering` orb and move the comparison of the `MARKETING_VERSION` that takes less than 30 sec in the `deploy-release` config so that the job can halt gracefully in case of any other amends being applied to the `Common.xcconfig` that is not meant to push a new release version.

### CODE REVIEW

File: .circleci/deploy-release.yml [hunk 1]
```diff
@@ -1,23 +1,31 @@
...
+      xcode: "14.3.1"
     resource_class: macos.x86.medium.gen2
     steps:
       - checkout
+      - run:
+          name: Check for MARKETING_VERSION change
+          command: |
+            git fetch origin $CIRCLE_BRANCH
+            git checkout $CIRCLE_BRANCH
+            CURRENT_VERSION=$(grep 'MARKETING_VERSION' Client/Configuration/Common.xcconfig | cut -d ' ' -f3)
```
Comment (ecotopian) [inline .circleci/deploy-release.yml:22]: out of curiosity: What's `cut -d ' ' -f3` exactly doing? I wouldn't mind a comment explaining it 😄

File: Client.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -12956,8 +12956,8 @@
 			isa = XCBuildConfiguration;
 			baseConfigurationReference = CA6FA5E529E6DA5E007595E4 /* EcosiaBetaDebug.xcconfig */;
 			buildSettings = {
-				CURRENT_PROJECT_VERSION = 78;
-				MARKETING_VERSION = 8.1.7;
+				CURRENT_PROJECT_VERSION = 0;
+				MARKETING_VERSION = "${inherited}";
```
Comment (ecotopian) [inline Client.xcodeproj/project.pbxproj:12960]: inheritance for the win 💰 😄

**Review by ecotopian — DISMISSED** [review]
Really awesome work! 🚀🚀🚀 No blockers at all, just a few minor comments. I am really fond to see this in action 💪

File: .circleci/deploy-release.yml (hunk 1 above)
Comment (d4r1091) [reply .circleci/deploy-release.yml:22]: Thanks for that @ecotopian ! 
Essentially, the `cut -d ' ' -f3` takes the output from `grep` command as input.
For example, let's assume that the `Common.xcconfig` file contains the following line:

`MARKETING_VERSION = 100.2.44`

`grep` will return -> `MARKETING_VERSION = 100.2.44`

The `cut` command will then extract the third field from the input, using a space (' ') as the delimiter.
Output: `100.2.44`

Comment (d4r1091) [reply .circleci/deploy-release.yml:22]: Can add a comment to it if you'd prefer 💪

Comment (d4r1091) [reply .circleci/deploy-release.yml:30]: Added 😉

**Review by lucaschifino — APPROVED** [review]
Amazing work ❤️ Really like the addition of checking for the marketing version before for efficiency 👏 

Also nice job with the very thorough PR description, even wonder if we should keep it somewhere not to "lose" it inside this PR 😅

**Review by ecotopian — APPROVED** [review]
Still great

**Pull request merged**